op-service/sources/eth_client.go

```diff
@@ -352,26 +315,31 @@ func (s *EthClient) PayloadByLabel(ctx context.Context, label eth.BlockLabel) (*
 // It verifies the receipt hash in the block header against the receipt hash of the fetched receipts
 // to ensure that the execution engine did not fail to return any receipts.
 func (s *EthClient) FetchReceipts(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Receipts, error) {
+	receipts, ok := s.recProvider.CachedReceipts(blockHash)
+	if ok {
+		// should always hit the cache
+		info, err := s.InfoByHash(ctx, blockHash)
```

Comment:
I feel like we should be leaving caching up to `InfoAndTxsByHash` and just calling that rather than adding this extra logic complexity.  We'd then be able to remove the `CachedReceipts` method and not leak the abstraction of the caching receipts provider.

If we need to fix the caching on `InfoAndTxsByHash` then that's a better option (or as you suggested somewhere, just use the tx hash list from the get block call).

Reply:
Good point, will remove the branch. In practice, we'd query headers, txs and receipts always together anyways, so it doesn't hurt to get it earlier. The only worthwhile optimization here would be to create a `InfoAndTxHashesByHash` function that only requires a header query, as I suggested in another comment.

Reply:
I remember now though why I added this. We always take the full list of transactions and derive the transaction hashes from it, which is quite the unnecessary additional work if we don't even fetch anything.

Reply:
I guess the assumption there is that it either would be in cache already (if the receipts are cached we must have requested the header too) or we'll likely need that header in the future anyway.

Reply:
What I meant was, further down in this method there's a `txHashes := eth.TransactionsToHashes(txs)`. I wanted to avoid recreating the tx hashes from full txs every time. But I realize now that tx hashes are cached inside every tx, so it's not too bad after all.

Reply:
removed it.